### PR TITLE
WIP: ansible_freeipa_module: Allow api_connect configuration.

### PR DESCRIPTION
**Pull request by rjeffman**
The function api_connect() is used to bootstrap the IPA API, and
configure the connection to the backend in this process. Currently,
only the `context` in which the commands will be executed, but other
attributes can also be set, like enabling/disabling LDAP cache layer.

By modifying ansible_freeipa_module.api_connect() to accept a set of
keyword arguments, it is possible to configure the bootstraping
process with the desired behavior.

The keyword arguments then have their type and values validated. Also,
if required, value mapping is applied, and, if not present, sensible
default values are provided.

If the argument type is invalid, or if an unexpected argument is given,
a TypeError is raised. If the value provided for the argument is not
in the list of accepted values a ValueError is raised.

Currently, the valid arguments are:
  - `context`: The context to run the module, accepts 'server',
    'client',or 'cli', the default is 'server', and 'client' is mapped
    to 'cli'.
  - `ldap_cache`: A bool value turning on/off the use LDAP caching
    layer for executing commands. Default is `True`.

### CODE REVIEW

File: plugins/module_utils/ansible_freeipa_module.py [hunk 1]
```diff
@@ -172,26 +172,57 @@ def temp_kdestroy(ccache_dir, ccache_name):
...
-            * `ansible-freeipa`
-            * `cli_installer`
+        context:
+            The context to execute IPA plugins. Can be any of
+            [server, client, cli].
+        ldap_cache:
+            Turn LDAP cache layer for commands on or off. Default is True.
         """
+        valid_args = {
+            "context": (str, {"client": "cil"}, ["server", "cli"], "server"),
```
Comment (t-woerner) [inline plugins/module_utils/ansible_freeipa_module.py:186]: What is `cil`?

Comment (t-woerner) [reply plugins/module_utils/ansible_freeipa_module.py:186]: It would also be nice to explain the structure of the dict.

File: tests/dnszone/test_dnszone.yml [hunk 2]
```diff
@@ -14,6 +14,7 @@
   - name: Ensure zone is present.
     ipadnszone:
       ipaadmin_password: SomeADMINpassword
+      ipaapi_ldap_cache: no
```
Comment (t-woerner) [inline tests/dnszone/test_dnszone.yml:17]: Why is ldap_cache disabled for this task?

File: plugins/modules/ipadnszone.py [hunk 3]
```diff
@@ -523,6 +523,7 @@ def get_argument_spec():
         ),
         ipaadmin_principal=dict(type="str", default="admin"),
         ipaadmin_password=dict(type="str", required=False, no_log=True),
+        ipaapi_ldap_cache=dict(type="bool", required=False),
```
Comment (t-woerner) [inline plugins/modules/ipadnszone.py:526]: `ipaapi_ldap_cache` has been added, but not `ipaapi_context`.

File: plugins/modules/ipavault.py [hunk 4]
```diff
@@ -742,7 +742,7 @@ def main():
             if ccache_name is not None:
                 os.environ["KRB5CCNAME"] = ccache_name
 
-        api_connect(context='ansible-freeipa')
+        api_connect(context='client')
```
Comment (t-woerner) [inline plugins/modules/ipavault.py:745]: Is this change related?

File: plugins/module_utils/ansible_freeipa_module.py (hunk 1 above)
Comment (rjeffman) [reply plugins/module_utils/ansible_freeipa_module.py:186]: `cil` is a typo.

I'll add a comment about `vaild_args`.

File: tests/dnszone/test_dnszone.yml (hunk 2 above)
Comment (rjeffman) [reply tests/dnszone/test_dnszone.yml:17]: Because by default, `ldap_cache` is disabled, and I wanted to test it changing the default behavior.

**Comment by rjeffman** [comment]
I'm closing this PR as it is covered by #631 and #645.
